### PR TITLE
Initial batch of bf updates

### PR DESCRIPTION
**Pull request by Quick-Flash**
Merging code from when bf 3.5.x started to august 25th 2018. Test flown on a whoop and all seems good, no real feature additions to speak of in here, except that I have completely ported over the bf 4.x matrix filter that Pawel was working on. Needs a few more tester verification. 

image

My worked stopped at the "Renamed 'fc/fc_' files to remove the double 'fc'" commit. More work to be done later.

### CODE REVIEW

**Comment by nerdCopter** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 11
           

Clones added
============
- src/main/target/SPRACINGF7DUAL/target.c  1
         

Clones removed
==============
+ src/main/target/AG3X/target.c  -1
+ src/main/target/OMNINXT/target.c  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604&bid=18205775)

File: src/main/drivers/accgyro/gyro_sync.c [hunk 1]
```diff
@@ -49,21 +49,40 @@ bool gyroSyncCheckUpdate(gyroDev_t *gyro)
...
+            gyroSamplePeriod = 31.25f;
+        } else {
+            switch (gyro->mpuDetectionResult.sensor) {
+            case BMI_160_SPI:
+                gyro->gyroRateKHz = GYRO_RATE_3200_Hz;
+                gyroSamplePeriod = 312.0f;
+                break;
+            case ICM_20649_SPI:
+                gyro->gyroRateKHz = GYRO_RATE_9_kHz;
+                gyroSamplePeriod = 1000000.0f / 9000.0f;
```
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:66]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/flight/gps_rescue.c [hunk 2]
```diff
@@ -417,24 +291,204 @@ void rescueAttainPosition()
...
+    }
+}
+
+static void sensorUpdate()
+{
+    rescueState.sensor.currentAltitudeCm = getEstimatedAltitudeCm();
+
+    // Calculate altitude velocity
+    static uint32_t previousTimeUs;
+    static int32_t previousAltitudeCm;
```
Comment (nerdCopter) [inline src/main/flight/gps_rescue.c:354]: Codacy Issue found: [The scope of the variable 'previousAltitudeCm' can be reduced.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c [hunk 3]
```diff
@@ -49,21 +49,40 @@ bool gyroSyncCheckUpdate(gyroDev_t *gyro)
...
-        gyro->gyroRateKHz = GYRO_RATE_3200_Hz;
-    } else if (gyro_use_32khz) {
-        //use full 32k
-        gyro->gyroRateKHz = GYRO_RATE_32_kHz;
+    float gyroSamplePeriod;
+
+    if (lpf != GYRO_HARDWARE_LPF_1KHZ_SAMPLE) {
+        if (gyro_use_32khz) {
+            gyro->gyroRateKHz = GYRO_RATE_32_kHz;
+            gyroSamplePeriod = 31.25f;
```
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:57]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/flight/gps_rescue.c [hunk 4]
```diff
@@ -51,15 +51,74 @@
...
+    uint8_t numSat;
+    float zVelocity; // Up/down movement in cm/s
+    float zVelocityAvg; // Up/down average in cm/s
+    float accMagnitude;
+    float accMagnitudeAvg;
+} rescueSensorData_s;
+
+typedef struct {
+    bool bumpDetection;
+    bool convergenceDetection;
```
Comment (nerdCopter) [inline src/main/flight/gps_rescue.c:102]: Codacy Issue found: [struct member 'rescueSanityFlags::convergenceDetection' is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c [hunk 5]
```diff
@@ -49,21 +49,40 @@ bool gyroSyncCheckUpdate(gyroDev_t *gyro)
...
+                gyro->gyroRateKHz = GYRO_RATE_8_kHz;
+                gyroSamplePeriod = 125.0f;
+                break;
+            }
+        }
+    } else {
+        switch (gyro->mpuDetectionResult.sensor) {
+        case ICM_20649_SPI:
+            gyro->gyroRateKHz = GYRO_RATE_1100_Hz;
+            gyroSamplePeriod = 1000000.0f / 1100.0f;
```
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:78]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/flight/gps_rescue.c [hunk 6]
```diff
@@ -394,7 +268,7 @@ void rescueAttainPosition()
     static float previousAltitudeError = 0;
     static int16_t altitudeIntegral = 0;
 
-    const int16_t altitudeError = (rescueState.intent.targetAltitude - rescueState.sensor.currentAltitude) / 100; // Error in meters
+    const int16_t altitudeError = (rescueState.intent.targetAltitudeCm - rescueState.sensor.currentAltitudeCm) / 100; // Error in meters
```
Comment (nerdCopter) [inline src/main/flight/gps_rescue.c:271]: Codacy Issue found: [Local variable 'altitudeError' shadows outer variable](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c [hunk 7]
```diff
@@ -49,21 +49,40 @@ bool gyroSyncCheckUpdate(gyroDev_t *gyro)
...
+                gyro->gyroRateKHz = GYRO_RATE_3200_Hz;
+                gyroSamplePeriod = 312.0f;
+                break;
+            case ICM_20649_SPI:
+                gyro->gyroRateKHz = GYRO_RATE_9_kHz;
+                gyroSamplePeriod = 1000000.0f / 9000.0f;
+                break;
+            default:
+                gyro->gyroRateKHz = GYRO_RATE_8_kHz;
+                gyroSamplePeriod = 125.0f;
```
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:70]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/flight/gps_rescue.c [hunk 8]
```diff
@@ -51,15 +51,74 @@
...
+    uint16_t groundSpeed;
+    uint8_t numSat;
+    float zVelocity; // Up/down movement in cm/s
+    float zVelocityAvg; // Up/down average in cm/s
+    float accMagnitude;
+    float accMagnitudeAvg;
+} rescueSensorData_s;
+
+typedef struct {
+    bool bumpDetection;
```
Comment (nerdCopter) [inline src/main/flight/gps_rescue.c:101]: Codacy Issue found: [struct member 'rescueSanityFlags::bumpDetection' is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c [hunk 9]
```diff
@@ -49,21 +49,40 @@ bool gyroSyncCheckUpdate(gyroDev_t *gyro)
...
+
+    if (lpf != GYRO_HARDWARE_LPF_1KHZ_SAMPLE) {
+        if (gyro_use_32khz) {
+            gyro->gyroRateKHz = GYRO_RATE_32_kHz;
+            gyroSamplePeriod = 31.25f;
+        } else {
+            switch (gyro->mpuDetectionResult.sensor) {
+            case BMI_160_SPI:
+                gyro->gyroRateKHz = GYRO_RATE_3200_Hz;
+                gyroSamplePeriod = 312.0f;
```
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:62]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/sensors/gyroanalyse.c [hunk 10]
```diff
@@ -245,54 +269,91 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state)
...
+            // get centerFreq in Hz from weighted bins
+            float centerFreq = dynNotchMaxHz;
+            float fftMeanIndex = 0;
             if (fftSum > 0) {
-                // idx was shifted by 1 to start at 1, not 0
-                fftMeanIndex = (fftWeightedSum / fftSum) - 1;
-                // the index points at the center frequency of each bin so index 0 is actually 16.125Hz
-                centerFreq = constrain(fftMeanIndex * fftResolution, dynNotchMinHz, dynNotchMaxCentreHz);
+                fftMeanIndex = (fftWeightedSum / fftSum);
+                centerFreq = fftMeanIndex * fftResolution;
```
Comment (nerdCopter) [inline src/main/sensors/gyroanalyse.c:329]: Codacy Issue found: [Redundant initialization for 'centerFreq'. The initialized value is overwritten before it is read.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c [hunk 11]
```diff
@@ -49,21 +49,40 @@ bool gyroSyncCheckUpdate(gyroDev_t *gyro)
...
+        }
+    } else {
+        switch (gyro->mpuDetectionResult.sensor) {
+        case ICM_20649_SPI:
+            gyro->gyroRateKHz = GYRO_RATE_1100_Hz;
+            gyroSamplePeriod = 1000000.0f / 1100.0f;
+            break;
+        default:
+            gyro->gyroRateKHz = GYRO_RATE_1_kHz;
+            gyroSamplePeriod = 1000.0f;
```
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:82]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

**Comment by nerdCopter** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 11
           

Clones added
============
- src/main/target/SPRACINGF7DUAL/target.c  1
         

Clones removed
==============
+ src/main/target/OMNINXT/target.c  -1
+ src/main/target/AG3X/target.c  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604&bid=18205775)

File: src/main/drivers/accgyro/gyro_sync.c (hunk 9 above)
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:62]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c (hunk 11 above)
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:82]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/flight/gps_rescue.c (hunk 8 above)
Comment (nerdCopter) [inline src/main/flight/gps_rescue.c:101]: Codacy Issue found: [struct member 'rescueSanityFlags::bumpDetection' is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c (hunk 3 above)
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:57]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/flight/gps_rescue.c (hunk 2 above)
Comment (nerdCopter) [inline src/main/flight/gps_rescue.c:354]: Codacy Issue found: [The scope of the variable 'previousAltitudeCm' can be reduced.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c (hunk 7 above)
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:70]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/flight/gps_rescue.c (hunk 6 above)
Comment (nerdCopter) [inline src/main/flight/gps_rescue.c:271]: Codacy Issue found: [Local variable 'altitudeError' shadows outer variable](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c (hunk 1 above)
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:66]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/flight/gps_rescue.c (hunk 4 above)
Comment (nerdCopter) [inline src/main/flight/gps_rescue.c:102]: Codacy Issue found: [struct member 'rescueSanityFlags::convergenceDetection' is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/sensors/gyroanalyse.c (hunk 10 above)
Comment (nerdCopter) [inline src/main/sensors/gyroanalyse.c:329]: Codacy Issue found: [Redundant initialization for 'centerFreq'. The initialized value is overwritten before it is read.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/drivers/accgyro/gyro_sync.c (hunk 5 above)
Comment (nerdCopter) [inline src/main/drivers/accgyro/gyro_sync.c:78]: Codacy Issue found: [Variable 'gyroSamplePeriod' is assigned a value that is never used.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)

File: src/main/sensors/gyroanalyse.c (hunk 10 above)
Comment (Quick-Flash) [reply src/main/sensors/gyroanalyse.c:329]: can someone besides codacy show me how this is redudant? i've looked for myself and was left confused.

File: src/main/drivers/accgyro/gyro_sync.c (hunk 5 above)
Comment (Quick-Flash) [reply src/main/drivers/accgyro/gyro_sync.c:78]: i understand it is useless, but i'm keeping it to make pulling more code from bf easier and once i pull more the issue will fix itself.

**Pull request closed without merging**